File: 백준/Gold/1759. 암호 만들기/암호 만들기.py
moum = ["a", "e", "i", "o", "u"]

L, C = map(int, input().split())
str_list = list(input().split())
str_list.sort()

# print(str_list)

# a..b...c..d..e 증가하는 순서로 넣어야 하고
# 최소 한개의 모음이 포함되어야 한다
temp = []
visited = [ False for _ in range(C) ]
result_list = [ 0 for _ in range(L) ]

def backtracking(depth, flag, idx):
    global result_list
    if(depth == L):
        count = 0
        for i in result_list :
            if (i in moum) :
                count += 1
        if(count >= 1 and L-count >= 2) :
            print(*result_list, sep='')
        return

    else :
        for i in range(idx, C) :
            if(flag & 1<<i) != 0 : continue
            result_list[depth] = str_list[i]
            backtracking(depth + 1, flag | (1<<i), i+1)

backtracking(0,0,0)

File: 백준/Gold/1759. 암호 만들기/암호 만들기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
	static char[] moum = {'a','e','i','o','u'};
	static int L, C;
	static ArrayList<Character> arrayList = new ArrayList<>();
	static ArrayList<Character> answer = new ArrayList<>();
	static boolean[] visited;
	static StringBuilder sb = new StringBuilder();
	
	public static void backtracking(int depth, int idx) {
		if(depth == L) {
			int count = 0;
			
			for(char i : answer) {
				for(int a=0; a<5; a++) {
					if(i == moum[a]) {
						count += 1;
					}
				}
			}
			
			// 최소 1개의 모음, 최소 두개의 자음	
			if(count >= 1 && (L-count) >= 2) {
				for(char i : answer) {
					sb.append(i);
				}
				sb.append("\n");
			}

			return;
		}
		
		
		for(int i=idx; i<C; i++) {
			if(!visited[i]) {
				visited[i] = true;
				answer.add(arrayList.get(i));
				backtracking(depth+1, i+1);
				answer.remove(answer.size()-1);
				visited[i] = false;
			}
		}
		
		
		
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer stk = new StringTokenizer(br.readLine());
		
		L = Integer.parseInt(stk.nextToken());
		C = Integer.parseInt(stk.nextToken());
		visited = new boolean[C+1];
		
		stk = new StringTokenizer(br.readLine());
		
		for(int i=0; i<C; i++) {
			arrayList.add(stk.nextToken().charAt(0));
		}
		
		Collections.sort(arrayList);
		
		backtracking(0,0);
		
		System.out.println(sb.toString());
	}
}